### ISSUE TITLE
Null-aware operator with cascade doesn't work with newlines?

### ISSUE THREAD

**Issue by mnordine** [dart-lang/language#1337]
Following code works:

```dart
void main() {
  int? x;
  x?..abs()..ceil()..floor();
}
```

This code gives an error:

```dart
void main() {
  int? x;
  x? // <- Conditions must have a static type of 'bool' - line 3
    ..abs()
    ..ceil()
    ..floor();
}
```

**Comment by mnordine** [comment]
Yeah, I understood that it caused parsing issues for you, I just wanted to point out the expectation and inconsistency from the programmers point of view. I also figured you knew about this and made a decision, but filed it just in case this hasn't come up before.

**Comment by lrhn** [comment]
I'm no sure we *couldn't* allow newlines between `?` and `..`. 
We'd probably have to declare that a `?` followed by a `..`, even with whitespace between them, is always going to be treated as a single `?..` operator. That would remove any ambiguity in parsing. I can't come up with an expression where it makes a difference, and if we ever introduce one, you'll have to parenthesize the part ending in `?`.

The slight difference between `!..` and `?..` can be confusing.
Unless we want to make writing the `?` on the previous line the *recommendation* (and update dartfmt), then it's not worth doing it, because dartfmt would just combine the `?..` anyway.

*If* we do it, we should probably also allow space inside `?.`, for consistency.

The currently mandated (by dartfmt) format is:
```dart
 longReceiver!
    .longMember1()!
    .longMember2();
 longReceiver!
    ..longMember1()
    ..longMember2();
 longReceiver
    ?.longMember1()
    ?.longMember2();
 longReceiver
    ?..longMember1()
    ..longMember2();
```

It's, at the very least, not consistent.
I kindof like the `?.`/`?..` starting the new line, it brings focus to the  operator, but I could live with `?` being trailing like `!` as well.
I would like them to *agree*.

Compare to:
```dart
 longReceiver!
    .longMember1()!
    .longMember2();
 longReceiver?
    .longMember1()?
    .longMember2();
 longReceiver!
    ..longMember1()
    ..longMember2();
 longReceiver?
    ..longMember1()
    ..longMember2();
```

or 

```dart
 longReceiver
    !.longMember1()
    !.longMember2();
 longReceiver
    ?.longMember1()
    ?.longMember2();
 longReceiver
    !..longMember1()
    ..longMember2();
 longReceiver
    ?..longMember1()
    ..longMember2();
```

To be honest, I think I prefer the `?` and `!` on the same line as a single `.` and on the receiver for cascade `..`.

**Comment by mnordine** [comment]
I didn't even realize it worked with putting the `?` on a newline. I guess I thought it looked so odd I didn't try it, but it works:

```dart
void main() {
  int? x;
  x
    ?..abs()
     ..ceil()
     ..floor();
}
```

**Comment by munificent** [comment]
The key difference here is that `!` is an expression and `?` is not. When you do:

```dart
foo!.bar();
```

You are performing two separate operations:

```dart
var bang = foo!;
bang.bar();
```

That is not the case with `?..`:

```dart
var question = foo?; // <-- Error, this is not a meaningful expression.
question..bar();
```

Allowing a newline in the middle of `?..` would be sort of like allowing a newline in the middle of `++`. It would send a signal to the reader that there are two distinct operations when there aren't. Always keeping `?..` together helps the reader understand that this is a single atomic operation.

> To be honest, I think I prefer the `?` and `!` on the same line as a single `.` and on the receiver for cascade ...

That was how I initially tried to format `!`, but it gets really weird in some cases:

```dart
var nonNullResult = nullableReceiver
    !.nullableMethod()
    !.anotherNullable()
    !.thirdNullable()!; // <-- :-O
```

Putting the `!` at the end of the previous line is consistent with how other postfix expressions like `()` and `[]` are formatted, and in fact can be combined with them:

```dart
verylongIdentifier![i]![j]
    .longIdentifier(i)![j]
    .aThird!;
```
